File: python/324_WiggleSortII.py
from unittest import TestCase
# https://leetcode.com/problems/wiggle-sort-ii/


def wiggled(nums):
    return not any(map(lambda i: nums[i] >= nums[i+1], range(0, len(nums)-1, 2)))


class WiggleSortII(TestCase):
    def wiggleSort(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        st = sorted(nums)
        n = len(nums)
        m = int((n+1)/2)
        l, r = st[:m][::-1], st[m:][::-1]
        for (i,v) in enumerate(l):
            nums[i*2] = v
        for (i,v) in enumerate(r):
            nums[i*2+1] = v

    def test1(self):
        nums = [1,2,3,4,5]
        self.wiggleSort(nums)
        self.assertTrue(wiggled(nums))

    def test2(self):
        nums = [4,5,5,6]
        self.wiggleSort(nums)
        self.assertTrue(wiggled(nums))

File: kotlin/src/main/kotlin/324.WiggleSortII.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/wiggle-sort-ii/


class WiggleSortII{
  fun wiggleSort(nums: IntArray): Unit {
    nums.sort()
    val n = nums.size
    val m = (n+1)/2
    val l = nums.slice(0 until m)
    val r = nums.slice(m until n)
    for((i,v) in l.reversed().withIndex())
      nums[i*2] = v
    for((i,v) in r.reversed().withIndex())
      nums[i*2+1] = v
  }

  fun wiggled(nums: IntArray) = !(0 until nums.size-1 step 2).any{ nums[it] >= nums[it+1] }

  @Test
  fun test1(){
    val nums = intArrayOf(1,2,3,4,5)
    wiggleSort(nums)
    assert(wiggled(nums))
  }

  @Test
  fun test2(){
    val nums = intArrayOf(4,5,5,6)
    wiggleSort(nums)
    assert(wiggled(nums))
  }

}